lib/bike_brigade/importers/mailchimp_importer.ex

```diff
@@ -7,83 +7,77 @@ defmodule BikeBrigade.Importers.MailchimpImporter do
   alias BikeBrigade.Importers.Importer
   alias BikeBrigade.Riders
   alias BikeBrigade.Riders.Rider
+  alias BikeBrigade.Messaging.SlackWebhook
 
   @count 100
   @mailchimp "mailchimp"
   @import_failed_tag %{name: "Dispatch Import Failed", status: "active"}
 
-  def sync_existing_rider(%Rider{} = rider) do
-    with {:ok, account} <- Mailchimp.Account.get(),
-         {:ok, list} <- Mailchimp.Account.get_list(account, get_config(:list_id)),
-         {:ok, member} <- Mailchimp.List.get_member(list, rider.email),
-         {:ok, rider_attrs} <- build_rider(member) do
-      Riders.update_rider(rider, rider_attrs)
-    end
-  end
-
-  def sync_riders do
-    Multi.new()
-    |> Multi.run(:get_last_synced, fn repo, _changes ->
+  # TODO make this a singleton? / add locking
+  def sync_riders(opts \\ nil) do
+    Repo.transaction(fn ->
       last_synced =
-        case repo.get_by(Importer, name: @mailchimp) do
+        # TODO add a row lock here.
+        case Repo.get_by(Importer, name: @mailchimp) do
           nil -> nil
           importer -> importer.data["last_synced"]
         end
 
-      {:ok, last_synced}
-    end)
-    |> Multi.run(:get_mailchimp_members, fn _repo, %{get_last_synced: last_synced} ->
-      get_members(last_synced)
-    end)
-    |> Multi.merge(fn %{get_mailchimp_members: mailchimp_members} ->
-      multi = Multi.new()
+      {:ok, members} = opts || get_members(last_synced)
```

Comment:
I think it's weird that we're passing `{:ok, [...]}` in tests

you can pass in the list of members itself, or better yet create an inner function that takes a list of members 

```elixir
def sync_riders() do
     Repo.transaction(fn -> 
        last_synced = ... # see comment above
        {:ok, members} = get_members(last_synced) # evaluate whether we need to return an `{:ok, members}` here - we needed the :ok tuples in the multi, but now we probably want to throw / revert the transaction if mailchimp is down
        process_members(members)
end

def process_members(members) do
      Repo.transaction(fn -> 
         ....
      end)
end
```

If you're asking yourself about the nested transactions, calling `Repo.transaction` inside another transaction is a no-op which is what we want -- see last paragraph https://hexdocs.pm/ecto/Ecto.Repo.html#c:transaction/2-use-with-function

Reply:
I'm not sure I'm following? We want to skip the `get_members` function in tests because it will get angry about a missing `api_key`. I feel like the better solution is to create a `FakeMailchimp` module (similar to `FakeSlack`). Thoughts? Maybe I am missing how you write the tests with the inner function.

Reply:
Oh sure, and have it be configurable to return some members.

I'm saying you could also just test the inner function